backend/app/controllers/spree/admin/products_controller.rb

```diff
@@ -30,6 +30,11 @@ def update
             param_attrs[:option_value_ids] = param_attrs[:option_value_ids].split(',')
           end
         end
+        if updating_variant_image_rules?
```

Comment:
I was hoping to have changes like that (as well as combining the ruleset models) in separate followup PRs. It's looking like a pretty big PR already and I think separate PRs would make it easier to review if everyone's OK with that.